### PR TITLE
Add Avro/AvroConfluent formats

### PR DESCRIPTION
**Pull request by oandrew**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):
Add Avro file input/output formats
Add AvroConfluent input format (for Kafka)

Detailed description (optional):
This adds support for Avro format input/output.
Schemas are dynamically matched/generated based on the table structure.

### Format `Avro` - input/output
Reading/writing  Avro data files.
Input example:
```
$ cat data.avro | avro-tools getschema /dev/stdin
{
  "type" : "record",
  "name" : "test",
  "fields" : [ {
    "name" : "a",
    "type" : {
      "type" : "array",
      "items" : {
        "type" : "array",
        "items" : "string"
      }
    }
  }, {
    "name" : "b",
    "type" : {
      "type" : "long",
      "logicalType" : "timestamp-millis"
    }
  }, {
    "name" : "c",
    "type" : [ "null", "long" ]
  } ]
}
$ cat avro.data | avrocat 
{"a": [["a1", "a2"], [""]], "b": 1578539189956, "c": null}
{"a": [], "b": 1578539189956, "c": {"long": 1}}
$ cat data.avro |  clickhouse-local   --input-format Avro --structure 'a Array(Array(String)), b DateTime64(3), c Nullable(Int64)' -q 'select *  from table format Pretty'
┏━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━┓
┃ a                  ┃                       b ┃    c ┃
┡━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━┩
│ [['a1','a2'],['']] │ 2020-01-08 21:06:29.956 │ ᴺᵁᴸᴸ │
├────────────────────┼─────────────────────────┼──────┤
│ []                 │ 2020-01-08 21:06:29.956 │    1 │
└────────────────────┴─────────────────────────┴──────┘
```
Output example:
```
$ clickhouse-local -q "select toNullable('str') as a, toInt64(123) as b, now64() as c  format Avro" | avro-tools getschema /dev/stdin
{
  "type" : "record",
  "name" : "row",
  "fields" : [ {
    "name" : "a",
    "type" : [ "null", "string" ]
  }, {
    "name" : "b",
    "type" : "long"
  }, {
    "name" : "c",
    "type" : {
      "type" : "long",
      "logicalType" : "timestamp-millis"
    }
  } ]
}
$ clickhouse-local -q "select toNullable('str') as a, toInt64(123) as b, now64() as c  format Avro" | avrocat
{"a": {"string": "str"}, "b": 123, "c": 1578539557810}
```

### Format `AvroConfluent` (experimental) - input
Support Confluent Platform Avro single-object framing format. It's often used with Kafka. Each Avro datum embeds a schema ID which is resolved   by looking it up in Schema Registry. This allows deserializing messages without knowing the schema in advance.  
`input_format_avro_schema_registry_url` setting allows to specify the Schema Registry URL.

Example:
```
CREATE TABLE data (`a` String, `b` Int64, `c` Nullable(Float64)) ENGINE = Kafka() SETTINGS kafka_broker_list = 'kafka-broker:9092', kafka_topic_list = 'topic', kafka_group_name = 'group', kafka_format = 'AvroConfluent', kafka_num_consumers = 1, input_format_avro_schema_registry_url = 'http://platform-schema-registry'
```

Notes:
- `AvroConfluent` only caches schemas per an instance of an InputFormat. This means SchemaRegistry will be queried each time a batch from Kafka is processed.

This PR also depends on some changes in  `contrib/boost` (adding iostreams) so for the purpose of getting some feedback I temporarily pointed it to my fork.

### Date Types matching:

| Avro data type (`INSERT`) | ClickHouse data type | Avro data type (`SELECT`) |
| -------------------- | ------------------ | ---- |
| `BOOL` | `UInt8` | `BOOL` |
| `INT` | `Int32` | `INT` |
| `LONG` | `Int64` | `LONG` |
| `FLOAT` | `Float32` | `FLOAT` |
| `DOUBLE` | `Float64` | `DOUBLE` |
| `STRING`, `BYTES`, `ENUM` | `String` | `STRING` |
| `FIXED(N)` | `FixedString(N)` | `FIXED(N)` |
| `ARRAY(T)` | `Array(T)` | `ARRAY(T)` |
| `UNION(NULL, T)` | `Nullable(T)` | `UNION(NULL, T)` |
| `ENUM` | `Enum8,Enum16` | `ENUM` |
| `INT(DATE)` | `Date` | `INT(DATE)` |
| `LONG(TIMESTAMP_MILLIS)` | `DateTime64(3)` | `LONG(TIMESTAMP_MILLIS)` |
| `LONG(TIMESTAMP_MICROS)` | `DateTime64(6)` | `LONG(TIMESTAMP_MICROS)` |

### CODE REVIEW

File: contrib/avro-cmake/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    ${AVROCPP_SOURCE_DIR}/parsing/JsonCodec.cc
+    ${AVROCPP_SOURCE_DIR}/parsing/ResolvingDecoder.cc
+    ${AVROCPP_SOURCE_DIR}/json/JsonIO.cc
+    ${AVROCPP_SOURCE_DIR}/json/JsonDom.cc
+    ${AVROCPP_SOURCE_DIR}/Resolver.cc
+    ${AVROCPP_SOURCE_DIR}/Validator.cc
+    )
+
+
+add_definitions(-std=c++17 -fPIC)
```
Comment (filimonov) [inline contrib/avro-cmake/CMakeLists.txt]: see https://github.com/ClickHouse/ClickHouse/pull/8464

**Comment by filimonov** [comment]
test?

**Comment by oandrew** [comment]
Updated PR with more information and added some usage examples.

**Comment by filimonov** [comment]
Nice 'Date Types matching:' table. Enum in input listed twice.

**Comment by oandrew** [comment]
> Nice 'Date Types matching:' table. Enum in input listed twice.

Yep, that's intentional -  Avro enum can be inserted into either String or Enum in Clickhouse.

**Comment by oandrew** [comment]
@alexey-milovidov could you take a look at this when you have a chance? Thanks

**Comment by alexey-milovidov** [comment]
Yes, give me a few days...

**Comment by filimonov** [comment]
> AvroConfluent only caches schemas per an instance of an InputFormat. This means SchemaRegistry will be queried each time a batch from Kafka is processed.

That sounds suboptimal. Especially for Kafka - it would be stupid to refetch schema for each block of data. Did you consider storing cached schemas somewhere else?

File: .gitmodules [hunk 2]
```diff
@@ -46,7 +46,7 @@
 	url = https://github.com/ClickHouse-Extras/protobuf.git
 [submodule "contrib/boost"]
 	path = contrib/boost
-	url = https://github.com/ClickHouse-Extras/boost.git
+	url = https://github.com/oandrew/clickhouse-boost
```
Comment (filimonov) [inline .gitmodules]: https://github.com/ClickHouse-Extras/boost/pull/2

Comment (filimonov) [reply .gitmodules]: remember to change it back...

**Comment by alexey-milovidov** [comment]
Doing HTTP request when parsing format is very strange and impose vulnerabilities.
Do you remember XXE vulnerabilities in XML parsers?
https://owasp.org/www-community/vulnerabilities/XML_External_Entity_(XXE)_Processing

I will check how we can ensure that `RemoteHostFilter` is available and used.

**Comment by alexey-milovidov** [comment]
And we don't setup any timeouts for HTTP request... You know - any request without timeout is anti-pattern.

**Pull request merged**